Fix billboards when clamped to ground with depthTestAgainstTerrain

hpinkos opened the pull request:
Fixes #2694

Before: 
![image](https://user-images.githubusercontent.com/3451886/40449462-7f0ed922-5ea6-11e8-8295-92f74499790d.png)

After:
![image](https://user-images.githubusercontent.com/3451886/40449465-82813488-5ea6-11e8-93a2-2fd25a26f7dc.png)

- This only changes billboards using `HeightReference.CLAMP_TO_GROUND`
- When the camera is greater than 2000m away from the billboard, use normal depth test against terrain
  - 2000 was picked arbitrarily through testing to see what looked reasonable
- When the camera is closer than 2000m to the billboard, check if the top left, top right and anchor points are behind terrain.  If all 3 points are behind terrain, hide the billboard.  Otherwise, show the billboard.
  - This can cause the billboard to show through terrain in some edge cases, but I think overall it's a vast improvement to what we have and brings us closer to enabling terrain by default.

Test code:
```js
var viewer = new Cesium.Viewer('cesiumContainer', {
    terrainProvider: Cesium.createWorldTerrain()
});
viewer.scene.globe.depthTestAgainstTerrain = true;

var e = viewer.entities.add({
    position : Cesium.Cartesian3.fromDegrees(-122.1958, 46.1915),
    billboard : {
        image : '../images/Cesium_Logo_overlay.png',
        heightReference : Cesium.HeightReference.CLAMP_TO_GROUND,
        verticalOrigin: Cesium.VerticalOrigin.BOTTOM
    }
});

viewer.trackedEntity = e;
```

hpinkos commented:
Thanks for helping me with this @bagnell!

What's the best way to write tests for this kind of thing?

bagnell commented:
I haven't seen the code yet, but I imagine you need to mock the globe so the clamp to ground just returns some constant height value. Also, see the [ground primitive tests](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Specs/Scene/GroundPrimitiveSpec.js#L86)
and classification tests. They change the primitive commands to execute during the globe pass which will be written into the globe depth texture.

hpinkos commented:
@bagnell I tried to write some tests, but everything breaks down trying to do a render test with height referenced billboards because it relies on the `QuadtreePrimitive` for adjusting the heights.  I'm guessing we don't have a good way to mock that and that's why we don't have any render tests for billboards using a height reference.

I think this is ready for review.  If you had other ideas for how to add tests let me know

bagnell commented on Source/Shaders/BillboardCollectionFS.glsl:
```diff
@@ -55,4 +64,39 @@ void main()
+#ifdef CLAMP_TO_GROUND
+    if (v_eyeDepth > -2000.0) {
+        vec2 adjustedST = v_textureCoordinates - v_textureOffset.xy;
+        adjustedST = adjustedST / (v_textureOffset.z - v_textureOffset.x, v_textureOffset.w - v_textureOffset.y);
```
I'm surprised this compiled. Add use the `vec2` constructor for the denominator.

hpinkos replied:
Thanks haha.  I hardly ever glsl so I'm not very familiar with the syntax

bagnell commented on Source/Shaders/BillboardCollectionFS.glsl:
```diff
@@ -55,4 +64,39 @@ void main()
+        vec2 st1 = ((v_dimensions.xy * (v_depthLookupTextureCoordinate1 - adjustedST)) + gl_FragCoord.xy) / czm_viewport.zw;
+        float logDepthOrDepth = czm_unpackDepth(texture2D(czm_globeDepthTexture, st1));
+        vec4 eyeCoordinate = czm_windowToEyeCoordinates(gl_FragCoord.xy, logDepthOrDepth);
+        float globeDepth1 = eyeCoordinate.z / eyeCoordinate.w;
```
Make this and the above two lines a function.

bagnell commented on Source/Shaders/BillboardCollectionFS.glsl:
```diff
@@ -55,4 +64,39 @@ void main()
+            eyeCoordinate = czm_windowToEyeCoordinates(gl_FragCoord.xy, logDepthOrDepth);
+            float globeDepth2 = eyeCoordinate.z / eyeCoordinate.w;
+
+            if (globeDepth2 > v_eyeDepth + czm_epsilon5)
```
One would think that all of these `if`s would increase performance for fragments that don't execute them, but, unfortunately, that's not the case. It would look cleaner if you compute all of the globe depths and then have one `if` combining all four conditionals for the discard.

bagnell commented on Source/Shaders/BillboardCollectionVS.glsl:
```diff
@@ -152,6 +166,14 @@ void main()
 
+#ifdef CLAMP_TO_GROUND
+    v_textureOffset = textureOffset;
+    v_depthLookupTextureCoordinate1 = (0.0, 0.9) - depthLookupST; //the origin
```
Are you sure these are right? Say the anchor point is the bottom right, then wouldn't you want to test `vec2(-1.0, 1.0)` instead of `vec2(1.0)`?

bagnell replied:
For the `vec*` constructors, a single argument duplicates it for all components.

hpinkos replied:
Yeah, you're right. I think I forgot to test different horizontal origins, this only works for center.

hpinkos commented:
@bagnell ready for another look.
Any ideas for what we can do for labels? I guess each letter is a separate billboard, which doesn't work well with this approach.

![label-ground](https://user-images.githubusercontent.com/3451886/40989701-a7c49648-68bc-11e8-9337-687046fc2b85.gif)

```js
var viewer = new Cesium.Viewer('cesiumContainer', {
    terrainProvider: Cesium.createWorldTerrain()
});
viewer.scene.globe.depthTestAgainstTerrain = true;

var e = viewer.entities.add({
    position : Cesium.Cartesian3.fromDegrees(-122.1958, 46.1915),
     label : {
        text : 'Clamped to ground',
        heightReference : Cesium.HeightReference.CLAMP_TO_GROUND,
        horizontalOrigin : Cesium.HorizontalOrigin.CENTER,
        verticalOrigin : Cesium.VerticalOrigin.BOTTOM,
        fillColor : Cesium.Color.BLACK,
        showBackground : true,
        backgroundColor : new Cesium.Color(1, 1, 1, 0.7),
        backgroundPadding : new Cesium.Cartesian2(8, 4)
    }
});

viewer.trackedEntity = e;
```

bagnell commented:
Look for `translate` in the vertex shader. It is the pixel offset to move each billboard in a label.

Did you test this with all of the other billboard options? You'll have to account for `pixelOffset`, `sizeInMeters`, `*ByDistance`, and others.

hpinkos commented:
@bagnell as far as I can tell, this is working correctly with things like `scale, `pixelOffset` `*ByDistance` and `sizeInMeters`.

The problem with the label is that each letter is checking the bounds of that letter instead of checking the bounds of the entire label.  That's why each letter flickers individually.

bagnell commented:
Can you add another attribute for the size of the whole label?

pjcozzi commented on Source/Scene/BillboardCollection.js:
```diff
@@ -302,7 +311,9 @@ define([
             BufferUsage.STATIC_DRAW, // PIXEL_OFFSET_SCALE_BY_DISTANCE_INDEX
-            BufferUsage.STATIC_DRAW  // DISTANCE_DISPLAY_CONDITION_INDEX
+            BufferUsage.STATIC_DRAW, // DISTANCE_DISPLAY_CONDITION_INDEX
+            BufferUsage.STATIC_DRAW,  // TEXTURE_OFFSET
```
Extra whitespace?

pjcozzi commented on Source/Scene/BillboardCollection.js:
```diff
@@ -736,6 +747,16 @@ define([
+            usage : buffersUsage[TEXTURE_OFFSET]
+        }, {
+            index : attributeLocations.dimensions,
+            componentsPerAttribute : 2,
```
Is there anyway to pack this into one component and perhaps store as `z` in `DISTANCE_DISPLAY_CONDITION_INDEX`?  Haven't looked closely, but perhaps precision requirements are not stiff here and fewer bits can be used?

hpinkos replied:
Is there a way to pack a `Cartesian2` into one float?  I'm not quite sure how to do this

pjcozzi replied:
@bagnell?

bagnell replied:
There are no bit-wise operators in GLSL but you can shift bits by multiplying powers of 2. See any of the `writeCompressedAttrib*` functions. So if you want to pack a `Cartesian2` into a float, you'll have 12 bits for each component. Is that enough precision? It would look something like this for packing:

```javascript
var LEFT_SHIFT12 = Math.pow(2.0, 12.0);
var width = Math.floor(Math.clamp(width, 0.0, LEFT_SHIFT12));
var height = Math.floor(Math.clamp(height, 0.0, LEFT_SHIFT12));
var dimensions = width * LEFT_SHIFT12 + height;
```

Then, reverse it in the vertex shader to unpack.

pjcozzi commented on Source/Scene/BillboardCollection.js:
```diff
@@ -1176,6 +1197,10 @@ define([
 
         var disableDepthTestDistance = billboard.disableDepthTestDistance;
+        if (billboard.heightReference === HeightReference.CLAMP_TO_GROUND && disableDepthTestDistance === 0.0) {
+            disableDepthTestDistance = 2000.0;
```
Should this be configurable at the scene level?  I think there are other billboard/label/depth-ish options there.

hpinkos replied:
It's configurable by setting `disableDepthTestDistance` to a different value

pjcozzi commented on Source/Shaders/BillboardCollectionVS.glsl:
```diff
@@ -10,11 +10,25 @@ attribute vec4 eyeOffset;                                  // eye offset in mete
 #endif
 
 varying vec2 v_textureCoordinates;
+#ifdef CLAMP_TO_GROUND
```
This is going to use a huge amount of memory bandwidth.

Looks easy to pack some varyings together.  Also perhaps some can be derived from others or more should be computed in the fragment shader if they are trivial?

pjcozzi replied:
Also, use more precise names - if possible - than `1`, `2`, and `3` for texture coordinates.

pjcozzi commented on Source/Shaders/BillboardCollectionVS.glsl:
```diff
@@ -10,11 +10,25 @@ attribute vec4 eyeOffset;                                  // eye offset in mete
 attribute vec4 pixelOffsetScaleByDistance;                 // near, nearScale, far, farScale
 attribute vec3 distanceDisplayConditionAndDisableDepth;    // near, far, disableDepthTestDistance
+#ifdef CLAMP_TO_GROUND
+attribute vec4 textureOffset;                              // the min and max x and y values for the texture coordinates
```
Should this be named something more precise like `textureCoordinatesBounds`?

pjcozzi commented on Source/Shaders/BillboardCollectionVS.glsl:
```diff
@@ -145,10 +162,34 @@ void main()
+    v_depthLookupTextureCoordinate1 = vec2(1.0) - depthOrigin; //the origin
+    if (v_depthLookupTextureCoordinate1.y == 1.0) //vertical origin is top
+    {
+        v_depthLookupTextureCoordinate2 = vec2(0.0, 0.0); //bottom left
```
Just `vec2(0.0)`

pjcozzi commented on Source/Shaders/BillboardCollectionFS.glsl:
```diff
@@ -40,5 +68,26 @@ void main()
+        float epsilonEyeDepth = v_eyeDepth + czm_epsilon5;
+
+        // negative values go into the screen
+        if (globeDepth1 > epsilonEyeDepth && globeDepth2 > epsilonEyeDepth && globeDepth3 > epsilonEyeDepth)
```
Make this a separate `if` statement after each call to `getGlobeDepth`.

On WebGL implementations with dynamic branches, it could exit early and avoid the memory bandwidth for the unnecessary texture reads.

Also, if not already done, order the calls to `getGlobeDepth` in the order of most likely to fail the depth test - I imagine origin first.

hpinkos replied:
I had it that way initially but @bagnell said it wouldn't work that way.  Is dynamic branches something that's only supported in certain browsers?

bagnell replied:
On a lot of GPUs with WebGL 1 support, there is no dynamic branches, but @pjcozzi is right. Early exits for GPUs with dynamic branches would be faster.

pjcozzi commented:
CHANGES.md.

hpinkos commented:
@bagnell this is ready for another review

bagnell commented:
You can get the background to disappear but not the glyphs. I used the code you posted above but changes the horizontal origin to be `LEFT` and the text is just `'Clamped'`.
![image](https://user-images.githubusercontent.com/1494815/41371962-f27c8458-6f19-11e8-9ca7-ce6694a0bc66.png)

bagnell commented on Source/Shaders/BillboardCollectionVS.glsl:
```diff
@@ -145,10 +160,36 @@ void main()
+    {
+        if (v_originTextureCoordinateAndTranslate.y == 0.0)
+        {
+            v_originTextureCoordinateAndTranslate.y = 0.1;
```
Why `0.1`?

bagnell commented on Source/Shaders/BillboardCollectionFS.glsl:
```diff
@@ -38,7 +73,32 @@ void main()
     czm_writeLogDepth();
+
+#ifdef CLAMP_TO_GROUND
+    if (v_eyeDepthAndDistance.x > -v_eyeDepthAndDistance.y) {
```
How much of a pain would it be to add this to both the vertex and fragment shader?

We would want it to execute in the vertex shader when `ContextLimits.maximumVertexTextureImageUnits > 0`; otherwise, execute in the fragment shader. That condition just means we can read from textures in the vertex shader. The reason we would want to do this is for performance. Instead of discarding the fragment, you would put the vertex behind the eye so that it gets clipped like we do for distance display conditions or translucency by distance.

hpinkos replied:
I'm going to do this in a separate PR because the math is different since there's no `gl_FragCoord` in the vertex shader.  I'll open an issue

bagnell commented:
The code looks good to me. The biggest thing would be executing that code in the vertex shader when possible.

Does anyone else want to test?

hpinkos commented:
> You can get the background to disappear but not the glyphs.

This was happening because of a slight discrepancy in the `eyeDepth` value of for the label glyphs and the background.  I fixed it by making the epsilon value I added to eyeDepth larger.  I'm not sure if that's the best solution, but the results were good

hpinkos commented:
@bagnell I think this is ready now

bagnell commented:
I can still get the background to disappear and some of the label glyphs disappear before others. This is with the vertical origin at the top and horizontal origin at he left.
![label_depth_test](https://user-images.githubusercontent.com/1494815/41563935-3e41d3f8-731f-11e8-9682-b40d684110d3.gif)

bagnell commented:
This crashes in IE since it doesn't have depth texture support.

hpinkos commented:
@bagnell thanks for catching that.  I changed IE so it has the behavior currently in master.  I also tested Firefox and Edge and they both work great.  And all the tests pass locally.

Anything else?

bagnell commented:
I still see the missing glyphs when the origin is left center. I also see all the glyphs disappear but the background is still visible when the origin is right center.

bagnell commented:
This is with the scale set to 1.5:
![label_depth_test_scale](https://user-images.githubusercontent.com/1494815/41625298-582d4532-73e6-11e8-85d3-390cce6e95ec.gif)

bagnell commented:
I'm seeing some strange behavior with rotated billboards. This has an origin at the bottom center and a rotation of pi/4.
![billboard_depth_test_rotation](https://user-images.githubusercontent.com/1494815/41625767-f577d93c-73e7-11e8-81c9-4b8bc0da3d4b.gif)

hpinkos commented:
Dagnabbit.  I thought I checked all the weird properties but I guess I missed some stuff 😓

hpinkos commented:
@bagnell I fixed `scale` and `rotation`/`alignedAxis`.  I'm a little stuck on `sizeInMeters` though

bagnell commented:
To fix `sizeInMeters`, multiply anything that has units in pixels with `czm_metersPerPixel(eyeCoordinates)`.

hpinkos commented:
@bagnell This should be ready for another look
I do still occasionally see blinking letters because of what I mentioned in this comment: https://github.com/AnalyticalGraphicsInc/cesium/pull/6621#issuecomment-398516724
The PR you submitted made it better, but it didn't fix it completely.  I'm going to play around with adding `floor` to different pixel calculations to see if that fixes it.  Right now the way we compute `translate` has all kinds of fractional values and I think things like that are causing the lookup to be 1px off every now and then.

But I believe I've fixed everything that was really wrong.  I tested a combination of different origins, scale, rotation and it seems to have pretty good behavior.

mramato commented:
> I'm going to play around with adding floor to different pixel calculations to see if that fixes it.

Be careful with this, clamping pixels can lead to very undesired behavior (especially when the text is moving).  That's the main reason we stopped clamping to screen space.  Not saying it applies here, just something to be aware of.

hpinkos commented:
> Be careful with this, clamping pixels can lead to very undesired behavior (especially when the text is moving)

@mramato thanks for noting that.  I should be able to make the location of the depth lookup consistent without effecting the position of the billboards.

hpinkos commented:
@bagnell nope, that didn't seem to help with the label letters.  Do you have any ideas?  I feel like it only happens if you're trying really hard and might not be a huge issue in normal application use.  Regardless, the behavior here is way better than what's in master now.

bagnell commented:
This looks good to me. I wasn't able to reproduce the missing letters at all where the last time I commented I could reproduce it reliably.

I'm merging this but don't forget about #6699.

esparano commented:
Is there a reason that this change only affects billboards with HeightReference.CLAMP_TO_GROUND? I'm having the same issues with billboards that are very close to (but not clamped to) the ground, so I had hoped that those would be fixed by this change.

On a side note, in my specific case, I was looking for a way to hide billboards based on ONLY the anchor point's visibility. Perhaps that could be an option in the future!

pjcozzi commented:
@esparano good point, this would also be useful for labels "on" 3D buildings, etc.

@hpinkos how hard is it to make this a separate boolean?  If easy, maybe do this soon; otherwise, submit an issue.  Should it default to true?  Is there ever a case when it should be false...hmm, might not need a flag at all if everyone likes the new depth testing...I dunno.
